jujugui/static/gui/src/test/test_app_renderer_extension.js

```diff
@@ -0,0 +1,145 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2013 Canonical Ltd.
+
+This program is free software: you can redistribute it and/or modify it under
+the terms of the GNU Affero General Public License version 3, as published by
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+describe('App Renderer Extension', function() {
+
+  var createElementStub,
+      juju,
+      Renderer,
+      renderer,
+      renderStub,
+      utils,
+      Y;
+
+  before(function(done) {
+    var requires = [
+      'base',
+      'app-renderer-extension',
+      'juju-tests-utils'
+    ];
+    Y = YUI(GlobalConfig).use(requires, function(Y) {
+      juju = Y.namespace('juju');
+      utils = Y.namespace('juju-tests.utils');
+      done();
+    });
+  });
+
+  beforeEach(function() {
+    Renderer = Y.Base.create(
+        'apprenderer', Y.Base, [juju.AppRenderer], {});
+    renderer = new Renderer();
+    // React method stubs.
+    renderStub = utils.makeStubMethod(ReactDOM, 'render');
+    this._cleanups.push(renderStub.reset);
+    createElementStub = utils.makeStubMethod(React, 'createElement');
+    this._cleanups.push(createElementStub.reset);
+    // Bootstrap various renderer attributes and properties.
+    var ecs = {
+      getCurrentChangeSet: utils.makeStubFunction({})
+    };
+    var Env = Y.Base.create('env', Y.Base, [], {
+      findFacadeVersion: utils.makeStubFunction('foo')
+    }, {
+      ATTRS: {
+        ecs: {}
+      }
+    });
+    renderer.env = new Env({ecs: ecs});
+    renderer.set('sandbox', false);
+    renderer._getAuth = utils.makeStubFunction({user: 'test'});
+    renderer.set('jujuEnvUUID', 'test');
+    renderer.state = {
+      getState: utils.makeStubFunction()
+    };
+    renderer.db = {
+      environment: {
+        set: utils.makeStubFunction()
+      }
+    };
+    renderer.changeState = utils.makeStubFunction();
+    renderer.showConnectingMask = utils.makeStubFunction();
+    renderer.switchEnv = utils.makeStubFunction();
+    renderer.createSocketURL = utils.makeStubFunction();
+  });
+
+  afterEach(function() {
+    renderer.destroy();
+  });
+
+  describe('_renderBreadcrumb', function() {
+    beforeEach(function() {
+      window.juju.components.HeaderBreadcrumb = {};
```

Comment:
These are the types of things which make this test suite unstable. Modifications to globals which aren't cleaned up afterwards.

You'll want to store whatever this value was before hand, and then reset it to that value in the afterEach